src/main/java/com/algorithmia/algorithmHandler/AlgorithmHandler.java

```diff
@@ -0,0 +1,107 @@
+package com.algorithmia.algorithmHandler;
+
+import java.lang.reflect.Method;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+public class AlgorithmHandler<INPUT, OUTPUT, STATE> {
+
+    private ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState;
+    private ReflectionHelper.DebuggableFunction<INPUT, OUTPUT> apply;
+    private Supplier<STATE> loadFunc;
+
+    private Class algorithmClass;
+    private STATE state;
+
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState, Supplier<STATE> loadFunc) {
+        this.applyWState = applyWState;
+        this.loadFunc = loadFunc;
+        this.algorithmClass = algorithmClass;
+    }
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState) {
+        this.applyWState = applyWState;
+        this.algorithmClass = algorithmClass;
+    }
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableFunction<INPUT, OUTPUT> apply) {
+        this.apply = apply;
+        this.algorithmClass = algorithmClass;
+    }
+
+    private void Load() {
+        if (this.loadFunc != null) {
+            state = this.loadFunc.get();
+            System.out.println("PIPE_INIT_COMPLETE");
+            System.out.flush();
+        }
+    }
+
+    private void ExecuteWithoutState(RequestHandler<INPUT> in, ResponseHandler out, Function<INPUT, OUTPUT> func) {
+        INPUT req = in.GetNextRequest();
+        while (req != null) {
+            OUTPUT output = func.apply(req);
+            out.writeToPipe(output);
+            req = in.GetNextRequest();
+        }
+    }
+
+    private void ExecuteWithState(RequestHandler<INPUT> in, ResponseHandler out, BiFunction<INPUT, STATE, OUTPUT> func) {
+        INPUT req = in.GetNextRequest();
+        while (req != null) {
+            OUTPUT output = func.apply(req, state);
+            out.writeToPipe(output);
+            req = in.GetNextRequest();
+        }
+    }
+
+    private void Execute(RequestHandler<INPUT> in, ResponseHandler out) {
+        if (this.applyWState != null && this.loadFunc != null) {
+            Load();
```

Comment:
we might want to wrap the call to Load in a try/catch in case errors happen there we should throw something with a clear message